s3-native-client/src/main/java/com/amazonaws/s3/S3NativeClient.java

```diff
@@ -17,41 +19,82 @@
     private final S3Client s3Client;
     private final String signingRegion;
 
-    public S3NativeClient(final String signingRegion,
-                          final ClientBootstrap clientBootstrap,
-                          final CredentialsProvider credentialsProvider,
-                          final long partSizeBytes,
-                          final double targetThroughputGbps) {
-        //keep signing region to construct Host header per-request
+    public S3NativeClient(final String signingRegion, final S3Client s3Client) {
         this.signingRegion = signingRegion;
-        final S3ClientOptions clientOptions = new S3ClientOptions()
-                .withClientBootstrap(clientBootstrap)
-                .withCredentialsProvider(credentialsProvider)
-                .withRegion(signingRegion)
-                .withPartSize(partSizeBytes)
+        this.s3Client = s3Client;
+    }
+
+    public S3NativeClient(final String signingRegion, final ClientBootstrap clientBootstrap,
+            final CredentialsProvider credentialsProvider, final long partSizeBytes,
+            final double targetThroughputGbps) {
+        // keep signing region to construct Host header per-request
+        this.signingRegion = signingRegion;
+        final S3ClientOptions clientOptions = new S3ClientOptions().withClientBootstrap(clientBootstrap)
+                .withCredentialsProvider(credentialsProvider).withRegion(signingRegion).withPartSize(partSizeBytes)
                 .withThroughputTargetGbps(targetThroughputGbps);
         s3Client = new S3Client(clientOptions);
     }
 
+    private void addCustomHeaders(List<HttpHeader> headers, HttpHeader[] customHeaders) {
+        assert headers != null : "Invalid argument - headers list is null";
+
+        if (customHeaders == null || customHeaders.length == 0) {
+            return;
+        }
+
+        for (HttpHeader customHeader : customHeaders) {
+
+            // Prevent duplicates and warn if any are found.
+            for (HttpHeader header : headers) {
+                if (customHeader.getName().equals(header.getName())) {
+                    Log.log(Log.LogLevel.Warn, Log.LogSubject.JavaCrtS3,
+                            "Custom header '" + customHeader.getName() + "' is overriding existing header.");
+                    headers.remove(header);
+                    break;
+                }
+            }
+
+            headers.add(customHeader);
+        }
+    }
+
+    private void addCustomQueryParameters(StringBuilder queryParameters, String customQueryParameters) {
+        assert queryParameters != null : "Invalid argument - query parameters is null";
+
+        if (customQueryParameters == null) {
+            return;
+        }
+
+        String trimmedCustomQueryParameters = customQueryParameters.trim();
+
+        if (trimmedCustomQueryParameters.equals("")) {
+            return;
+        }
+
+        queryParameters.append("?" + trimmedCustomQueryParameters);
```

Comment:
If it's assumed(required) that queryParameters comes in empty then maybe it should be created and returned instead.